synapse/federation/federation_client.py

```diff
@@ -386,51 +383,75 @@ async def _check_sigs_and_hash_and_fetch(
         Returns:
             A list of PDUs that have valid signatures and hashes.
         """
-        deferreds = self._check_sigs_and_hashes(room_version, pdus)
+        valid_pdus = await yieldable_gather_results(
+            self._check_sigs_and_hash_and_fetch_one,
+            pdus,
+            origin=origin,
+            room_version=room_version,
+            outlier=outlier,
+        )
```

Comment:
does this need a concurrency limit?

Reply:
Ah, yeah, may as well actually. I've left it so we still do the limiting manually in the `send_join` function, as that we we don't need to convert from list to map after the fact.